src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.ClientCertProvider.cs

```diff
@@ -0,0 +1,255 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography.X509Certificates;
+using System.Security.Cryptography;
+using Microsoft.Win32.SafeHandles;
+
+using CURLoption = Interop.libcurl.CURLoption;
+using CURLcode  = Interop.libcurl.CURLcode;
+using CURLINFO = Interop.libcurl.CURLINFO;
+namespace System.Net.Http
+{
+    internal partial class CurlHandler : HttpMessageHandler
+    {
+
+        private sealed class ClientCertProvider : IDisposable
+        {
+            private static readonly Interop.libcurl.ssl_ctx_callback s_sslCallback = CurlSslCallback;
+            private string _certificateFile;
+            private readonly X509Certificate2 _certificate;
+
+            /* resources relevant for SSL callbacks */
+            private readonly RSAOpenSsl _rsa;
+            private readonly GCHandle _gcHandle;
+
+            private ClientCertProvider(X509Certificate2 certificate)
+            {
+                Debug.Assert(certificate != null, "Certificate created cannot be null");
+                _certificate = certificate;
+                _rsa = certificate.GetRSAPrivateKey() as RSAOpenSsl;
```

Comment:
You shouldn't need to hold on to the RSA object; DuplicateKeyHandle() returns an object with an independent lifetime.  That's what you should be persisting and later disposing..

The RSAOpenSsl you should just dispose in this method after asking it for a key handle.